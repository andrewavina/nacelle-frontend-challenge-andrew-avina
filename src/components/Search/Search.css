.search-container {
  width: 400px;
  margin: 20px auto;
  font-family: Arial, sans-serif;
}

.search-input-container {
  display: flex;
  align-items: center;
  margin-bottom: 20px;
}

.search-input {
  flex: 1;
  padding: 10px;
  font-size: 16px;
  border: 1px solid #ccc;
  border-radius: 4px;
}

.clear-button {
  margin-left: 10px;
  padding: 10px 20px;
  font-size: 14px;
  background-color: #f44336;
  color: white;
  border: none;
  border-radius: 4px;
  cursor: pointer;
}

.clear-button:hover {
  background-color: #d32f2f;
}

.loading-state {
  color: #3f51b5;
  font-size: 14px;
}

.error-state {
  color: #f44336;
  font-size: 14px;
}

.results-list {
  list-style: none;
  padding: 0;
  margin: 0;
}

.result-item {
  background-color: #f5f5f5;
  padding: 10px;
  margin-bottom: 5px;
  border-radius: 4px;
  cursor: pointer;
}

.result-item:hover {
  background-color: #e0e0e0;
}

.sr-only {
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  border: 0;
}
