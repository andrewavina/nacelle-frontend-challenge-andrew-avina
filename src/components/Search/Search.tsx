import React, { useState } from "react";
import "./Search.css";

interface SearchProps {
  onSearch: (query: string) => void;
  placeholder?: string;
}

const sampleData = [
  { id: "1", title: "React Basics" },
  { id: "2", title: "TypeScript Tutorial" },
  { id: "3", title: "Frontend Testing" },
  { id: "4", title: "Redux Guide" },
  { id: "5", title: "React Hooks Demo" },
];

const Search: React.FC<SearchProps> = ({
  onSearch,
  placeholder = "Search...",
}) => {
  const [query, setQuery] = useState("");
  const [results, setResults] = useState<typeof sampleData | null>(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState("");

  const handleSearch = (value: string) => {
    setQuery(value);
    setLoading(true);
    setError("");
    setResults(null);

    // Simulating an API call delay here to show the loading state per requirement
    setTimeout(() => {
      try {
        const filteredData = sampleData.filter((item) =>
          item.title.toLowerCase().includes(value.toLowerCase()),
        );
        setResults(filteredData);
        if (filteredData.length === 0) {
          setError("No results found.");
        }
      } catch {
        setError("Something went wrong.");
      } finally {
        setLoading(false);
        onSearch(value);
      }
    }, 500);
  };

  const clearSearch = () => {
    setQuery("");
    setResults(null);
    setError("");
    onSearch("");
  };

  return (
    <div className="search-container">
      <div className="search-input-container">
        <label htmlFor="search-input" className="sr-only">
          Search
        </label>
        <input
          className="search-input"
          id="search-input"
          type="text"
          value={query}
          onChange={(e) => handleSearch(e.target.value)}
          placeholder={placeholder}
          aria-describedby="search-error"
        />
        {query && (
          <button
            className="clear-button"
            aria-label="Clear search"
            onClick={clearSearch}
          >
            Clear
          </button>
        )}
      </div>

      {loading && <p className="loading-state">Loading...</p>}

      {error && !loading && (
        <p className="error-state" id="search-error" role="alert">
          {error}
        </p>
      )}

      {results && !loading && (
        <ul className="results-list">
          {results.map((result) => (
            <li key={result.id} className="result-item" tabIndex={0}>
              {result.title}
            </li>
          ))}
        </ul>
      )}
    </div>
  );
};

export default Search;
