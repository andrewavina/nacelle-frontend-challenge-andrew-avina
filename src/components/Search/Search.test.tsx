import { render, fireEvent, screen } from "@testing-library/react";
import Search from "./Search";

describe("Search component unit tests:", () => {
  test("input handling and displaying results:", () => {
    const { getByPlaceholderText, getByText } = render(
      <Search onSearch={jest.fn()} placeholder="Search..." />,
    );

    const input = getByPlaceholderText("Search...");
    fireEvent.change(input, { target: { value: "React" } });
    expect(getByText("Loading...")).toBeInTheDocument();
  });

  test("clear button clears input and results", () => {
    const { getByPlaceholderText, getByText } = render(
      <Search onSearch={jest.fn()} placeholder="Search..." />,
    );

    const input = getByPlaceholderText("Search...");
    fireEvent.change(input, { target: { value: "React" } });
    fireEvent.click(getByText("Clear"));
    expect(input).toHaveValue("");
  });

  test("displays no results message", async () => {
    const { getByPlaceholderText, findByText } = render(
      <Search onSearch={jest.fn()} placeholder="Search..." />,
    );

    fireEvent.change(getByPlaceholderText("Search..."), {
      target: { value: "Unknown" },
    });
    expect(await findByText("No results found.")).toBeInTheDocument();
  });
});

describe("Search component snapshot tests:", () => {
  test("matches snapshot for initial render", () => {
    const { asFragment } = render(
      <Search onSearch={() => {}} placeholder="Search for tutorials..." />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
