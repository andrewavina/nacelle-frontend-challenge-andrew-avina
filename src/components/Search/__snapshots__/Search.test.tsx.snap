// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component snapshot tests: matches snapshot for initial render 1`] = `
<DocumentFragment>
  <div
    class="search-container"
  >
    <div
      class="search-input-container"
    >
      <label
        class="sr-only"
        for="search-input"
      >
        Search
      </label>
      <input
        aria-describedby="search-error"
        class="search-input"
        id="search-input"
        placeholder="Search for tutorials..."
        type="text"
        value=""
      />
    </div>
  </div>
</DocumentFragment>
`;
